torch/_inductor/compile_fx.py

```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+    graph or tensor which is an output of a graph remains alive.
+    """
+
+    def __init__(self):
+        self.tape_manager = None
+        self.alive_cudagraphify_objs = 0
+
+        # only set in the case where the CudaGraphify objects die while their
+        # tensor outputs remain live
+        self.live_tensors_count = 0
+        # Used to keep alive the memory pool if the CudaGraph object dies
+        self.graph = None
+        self.lock = threading.Lock()
+
+    def finalize_tensor(self):
+        with self.lock:
+            self.live_tensors_count -= 1
+            if self.live_tensors_count == 0:
+                self.tape_manager = None
+                self.graph = None
+
+    def finalize_cuda_graphify(self, graph: torch.cuda.CUDAGraph, tape_manager):
+        with self.lock:
+            self.alive_cudagraphify_objs -= 1
+            if self.alive_cudagraphify_objs != 0:
+                return
+            live_tensors = tape_manager.live_cuda_graph_managed_tensors()
+            if not live_tensors:
+                self.tape_manager = None
+                return
+
+            self.graph = graph
+            for t in live_tensors:
+                self.live_tensors_count += 1
+                weakref.finalize(t, self.finalize_tensor)
+
+    def get_tape_manager(self, obj: Optional["CudaGraphify"] = None):
+        # if the live tensors count is set, then the memory pool is still
+        # being used even if the tape manager and CudaGraphTape objects are dead
+        if self.live_tensors_count:
+            return None
+
+        if self.tape_manager is None or self.tape_manager() is None:
+            tape_manager = CUDAGraphTapeManger()
+            self.tape_manager = weakref.ref(tape_manager)
+        else:
+            tape_manager = self.tape_manager()
+
+        if isinstance(obj, CudaGraphify):
+            with self.lock:
+                self.alive_cudagraphify_objs += 1
+
+            # needs to be weak refs otherwise finalizer would keep obj alive
+            obj_ref = weakref.ref(obj)
+            finalize_fn = functools.partial(
+                self.finalize_cuda_graphify, graph=obj.graph, tape_manager=tape_manager
+            )
+            weakref.finalize(obj, finalize_fn)
+
+        return tape_manager
+
+
+# TODO - actually make thread local - need to register at TLS that gets copied over
+# in aten/ThreadLocalState
+tape_manager_container = TapeManagerContainer()
+
+
+class CudaGraphify(object):
+    def __init__(self, model, inputs, static_input_idxs=()):
+
+        assert isinstance(inputs, (list, tuple))
+        self.graph = torch.cuda.CUDAGraph()
+        self.model = model
+        static_input_idxs = remove_unaligned_input_idxs(inputs, static_input_idxs)
+
+        tape_manager = tape_manager_container.get_tape_manager(self)
+        tape_manager is not None
+        self.tape_manager = tape_manager
```

Comment:
This also looks reference-cycle-y. What's the overall ownership structure for tapes and tape managers and cudagraphify objects?